libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+    )
+  let pubKeyBytes = pubKeyBytesResult.get()
+
+  # Generate the SignedKey ASN.1 structure
+  return generateSignedKey(signature, pubKeyBytes)
+
+proc generate*(
+    identityKeyPair: KeyPair
+): (seq[byte], seq[byte]) {.
+    raises: [
+      KeyGenerationError, CertificateCreationError, ASN1EncodingError,
+      IdentityPubKeySerializationError, IdentitySigningError, TLSCertificateError,
+    ]
+.} =
+  ## Generates a self-signed X.509 certificate with the libp2p extension.
+  ##
+  ## Parameters:
+  ## - `identityKeyPair`: The peer's identity key pair.
+  ##
+  ## Returns:
+  ## A tuple containing:
+  ## - The DER-encoded certificate.
+  ## - The DER-encoded private key.
+  ##
+  ## Raises:
+  ## - `KeyGenerationError` if key generation fails.
+  ## - `CertificateCreationError` if certificate creation fails.
+  ## - `ASN1EncodingError` if encoding fails.
+  # Initialize entropy and DRBG contexts
+  var
+    entropy = mbedtls_entropy_context()
+    ctrDrbg = mbedtls_ctr_drbg_context()
+    crt: mbedtls_x509write_cert
+    certKey: mbedtls_pk_context
+    ret: cint
+
+  mbedtls_entropy_init(addr entropy)
+  mbedtls_ctr_drbg_init(addr ctrDrbg)
+  mbedtls_x509write_crt_init(addr crt)
+  mbedtls_pk_init(addr certKey)
+
+  defer:
+    mbedtls_entropy_free(addr entropy)
+    mbedtls_ctr_drbg_free(addr ctrDrbg)
+    mbedtls_pk_free(addr certKey)
+    mbedtls_x509write_crt_free(addr crt)
+
+  # Seed the random number generator
+  let personalization = "libp2p_tls"
+  ret = mbedtls_ctr_drbg_seed(
+    addr ctrDrbg,
+    mbedtls_entropy_func,
+    addr entropy,
+    cast[ptr byte](personalization.cstring),
+    personalization.len.uint,
+  )
+  if ret != 0:
+    raise newException(KeyGenerationError, "Failed to seed CTR_DRBG")
+
+  # Initialize certificate key
+  ret = mbedtls_pk_setup(addr certKey, mbedtls_pk_info_from_type(MBEDTLS_PK_ECKEY))
+  if ret != 0:
+    raise newException(KeyGenerationError, "Failed to set up certificate key context")
+
+  # Generate key pair for the certificate
+  let G =
+    try:
+      mb_pk_ec(certKey)
+    except MbedTLSError as e:
+      raise newException(KeyGenerationError, e.msg)
+  ret = mbedtls_ecp_gen_key(EC_GROUP_ID, G, mbedtls_ctr_drbg_random, addr ctrDrbg)
+  if ret != 0:
+    raise
+      newException(KeyGenerationError, "Failed to generate EC key pair for certificate")
+
+  ## Initialize libp2p extension
+  let libp2pExtension = makeLibp2pExtension(identityKeyPair, certKey)
+
+  # Set the Subject and Issuer Name (self-signed)
+  ret = mbedtls_x509write_crt_set_subject_name(addr crt, "CN=libp2p.io")
+  if ret != 0:
+    raise newException(CertificateCreationError, "Failed to set subject name")
+
+  ret = mbedtls_x509write_crt_set_issuer_name(addr crt, "CN=libp2p.io")
+  if ret != 0:
+    raise newException(CertificateCreationError, "Failed to set issuer name")
+
+  # Set Validity Period
+  let notBefore = "19750101000000"
+  let notAfter = "40960101000000"
```

Comment:
Is it from the spec?

Reply:
They don't mention this in the spec, I used the same values in the Rust implementation.